Source/WebCore/platform/graphics/cg/PathCG.cpp

```diff
@@ -21,167 +21,316 @@
  * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
  * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
 #include "config.h"
-#include "Path.h"
+#include "PathCG.h"
 
 #if USE(CG)
 
-#include "AffineTransform.h"
-#include "FloatRect.h"
 #include "GraphicsContextCG.h"
-#include "IntRect.h"
-#include <pal/spi/cg/CoreGraphicsSPI.h>
-#include <wtf/MathExtras.h>
+#include <wtf/NeverDestroyed.h>
 #include <wtf/RetainPtr.h>
-#include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
-static size_t putBytesNowhere(void*, const void*, size_t count)
+std::unique_ptr<PathCG> PathCG::create()
 {
-    return count;
+    return makeUnique<PathCG>();
 }
 
-static RetainPtr<CGContextRef> createScratchContext()
+std::unique_ptr<PathCG> PathCG::create(const PathStream& stream)
 {
-    CGDataConsumerCallbacks callbacks = { putBytesNowhere, 0 };
-    auto consumer = adoptCF(CGDataConsumerCreate(0, &callbacks));
-    auto context = adoptCF(CGPDFContextCreate(consumer.get(), 0, 0));
+    auto pathCG = PathCG::create();
+    if (!pathCG)
+        return nullptr;
 
-    CGFloat black[4] = { 0, 0, 0, 1 };
-    CGContextSetFillColor(context.get(), black);
-    CGContextSetStrokeColor(context.get(), black);
+    stream.applySegments([&](const PathSegment& segment) {
+        pathCG->appendSegment(segment);
+    });
 
-    return context;
+    return pathCG;
 }
 
-static inline CGContextRef scratchContext()
+std::unique_ptr<PathCG> PathCG::create(RetainPtr<CGMutablePathRef>&& platformPath)
 {
-    static NeverDestroyed<RetainPtr<CGContextRef>> context = createScratchContext();
-    return context.get().get();
+    return makeUnique<PathCG>(WTFMove(platformPath));
 }
 
-Path Path::polygonPathFromPoints(const Vector<FloatPoint>& points)
+PathCG::PathCG()
+    : m_platformPath(adoptCF(CGPathCreateMutable()))
 {
-    Path path;
-    if (points.size() < 2)
-        return path;
+}
 
-    Vector<CGPoint, 32> cgPoints;
-    cgPoints.reserveInitialCapacity(points.size());
-    for (size_t i = 0; i < points.size(); ++i)
-        cgPoints.uncheckedAppend(points[i]);
+PathCG::PathCG(RetainPtr<CGMutablePathRef>&& platformPath)
+    : m_platformPath(WTFMove(platformPath))
+{
+    ASSERT(m_platformPath);
+}
 
-    CGPathAddLines(path.ensurePlatformPath(), nullptr, cgPoints.data(), cgPoints.size());
-    path.closeSubpath();
-    return path;
+std::unique_ptr<PathImpl> PathCG::clone() const
+{
+    return create({ platformPath() });
 }
 
-void Path::createCGPath() const
+PlatformPathPtr PathCG::platformPath() const
 {
-    if (m_path)
-        return;
+    return m_platformPath.get();
+}
+
+PlatformPathPtr PathCG::ensurePlatformPath()
```

Comment:
I think this should have a different name. `ensurePlatformPath` sounds like it just creates one if one doesn't already exist, but actually it's about ensuring we have a unique one.

Reply:
It is renamed to ensureMutablePlatformPath() since the platform path is single writer multiple readers. And this function is only called when we want to modify the platform path.